hbase-endpoint/src/test/java/org/apache/hadoop/hbase/coprocessor/TestAsyncTableCoprocessorEndpoint.java

```diff
@@ -0,0 +1,444 @@
+    futureAndCallback.getFirst().join();
+
+    RegionInfo regionToSplit = TEST_UTIL.getAdmin().getRegions(TABLE_NAME).stream()
+      .filter(
+        regionInfo -> regionInfo.getStartKey().length > 0 && regionInfo.getEndKey().length > 0)
+      .findFirst().orElseThrow(IllegalArgumentException::new);
+    TEST_UTIL.getAdmin().splitRegionAsync(regionToSplit.getRegionName(),
+      Bytes.split(regionToSplit.getStartKey(), regionToSplit.getEndKey(), 1)[1]).get();
+
+    futureAndCallback = buildCallbackForAllRegions();
+    table
+      .coprocessorService(ColumnAggregationService::newStub, sum(), futureAndCallback.getSecond())
+      .execute();
+
+    Map<String, List<Pair<SumResponse, Throwable>>> responsesByRegion =
+      futureAndCallback.getFirst().get(10, TimeUnit.SECONDS);
+    assertResultsAreCorrect(responsesByRegion);
+  }
+
+  @Test
+  public void itCallsEndpointOnRegionsWithSplitAndMerge() throws Exception {
+    AsyncTable<?> table = getTable.get();
+
+    // call the endpoints before to ensure the meta cache has the old region locations
+    Pair<CompletableFuture<Map<String, List<Pair<SumResponse, Throwable>>>>,
+      PartialResultCoprocessorCallback<ColumnAggregationService, SumResponse>> futureAndCallback =
+        buildCallbackForAllRegions();
+    table
+      .coprocessorService(ColumnAggregationService::newStub, sum(), futureAndCallback.getSecond())
+      .execute();
+    futureAndCallback.getFirst().join();
+
+    // it takes a while for a merged region to become splitable, so simulate it by recreating the
+    // table with new splits
+    TEST_UTIL.deleteTable(TABLE_NAME);
+    createTestTable(1);
+    RegionInfo regionToSplit = TEST_UTIL.getAdmin().getRegions(TABLE_NAME).stream().findFirst()
+      .orElseThrow(IllegalArgumentException::new);
+    TEST_UTIL.getAdmin().splitRegionAsync(regionToSplit.getRegionName(), Bytes.fromHex("053333"))
+      .get(); // use a new split point that wasn't in the original table
+
+    futureAndCallback = buildCallbackForAllRegions();
+    table
+      .coprocessorService(ColumnAggregationService::newStub, sum(), futureAndCallback.getSecond())
+      .execute();
+
+    Map<String, List<Pair<SumResponse, Throwable>>> responsesByRegion =
+      futureAndCallback.getFirst().get(10, TimeUnit.SECONDS);
+    assertResultsAreCorrect(responsesByRegion);
+  }
+
+  @Test
+  public void itCallsEndpointDuringRegionSplitProcedure() throws Exception {
+    AsyncTable<?> table = getTable.get();
+
+    RegionInfo regionToSplit = TEST_UTIL.getAdmin().getRegions(TABLE_NAME).stream()
+      .filter(
+        regionInfo -> regionInfo.getStartKey().length > 0 && regionInfo.getEndKey().length > 0)
+      .findFirst().orElseThrow(IllegalArgumentException::new);
+
+    Pair<CompletableFuture<Map<String, List<Pair<SumResponse, Throwable>>>>,
+      PartialResultCoprocessorCallback<ColumnAggregationService, SumResponse>> futureAndCallback =
+        buildCallbackForAllRegions((region, prevRequestNumber) -> {
```

Comment:
This splits the region in between the first and second RPC for that region; in this case, because the callback was already called for the first RPC, we fail the region